### PR TITLE
API: Modify the descriptor struct ABI (to some degree)

### PR DESCRIPTION
**Pull request by seberg**
This a WIP, changes are self contained up to now (first 3 commits, I am not planning to rebase them away):
* Bump elsize to `npy_intp` and flags to `uint64` (I don't mind changing the flags, e.g. to uint32).  And a small fix for pickling due to changed flags (may need more fixes to pickle huge dtypes).
  * Also reorders the later struct fields.
* Creates a new `PyArray_DescrProto` struct that is exactly like the old one and modify the legate dtype registration to use it.

This is currently a bit tedious, but overall a surprisingly small change, it of course breaks ABI, so the next steps are:
1. Create run-time dependent macro's to access fields and rename `elsize` to `itemsize`.
2. Hide fields which are limited to our legacy dtypes into another custom struct extending the public `PyArray_Descr` struct.  (I would prefer not blessing some fields for user dtypes.  It has the added advantage that our main structure needs to be smaller or equal the old one for Cython.)

After the two things above are done, this can be mergable.  We then need to discuss modifying the struct further.  E.g. @ngoldbaum and I would like to be able to represent bitsize (maybe an additional field or a bitsized flag which NumPy would enforce not being set for NumPy arrays:  A bitsized dtype would need a way to pad it to full bytes to be attached to a NumPy array.).

### CODE REVIEW

File: numpy/_core/src/multiarray/buffer.c [hunk 1]
```diff
@@ -398,22 +398,22 @@ _buffer_format_string(PyArray_Descr *descr, _tmp_string_t *str,
...
+            PyOS_snprintf(buf, sizeof(buf), "%" NPY_INTP_FMT "s", descr->elsize);
             if (_append_str(str, buf) < 0) return -1;
             break;
         }
         case NPY_UNICODE: {
             /* NumPy Unicode is always 4-byte */
             char buf[128];
             assert(descr->elsize % 4 == 0);
-            PyOS_snprintf(buf, sizeof(buf), "%dw", descr->elsize / 4);
+            PyOS_snprintf(buf, sizeof(buf), "%" NPY_INTP_FMT"w", descr->elsize / 4);
```
Comment (mhvk) [inline numpy/_core/src/multiarray/buffer.c:409]: Missing space after _FMT (for consistency)

File: numpy/_core/src/multiarray/descriptor.c [hunk 2]
```diff
@@ -2028,11 +2028,15 @@ static PyMemberDef arraydescr_members[] = {
...
+        T_PYSSIZET, offsetof(PyArray_Descr, elsize), READONLY, NULL},
     {"alignment",
-        T_INT, offsetof(PyArray_Descr, alignment), READONLY, NULL},
+        T_PYSSIZET, offsetof(PyArray_Descr, alignment), READONLY, NULL},
     {"flags",
-        T_BYTE, offsetof(PyArray_Descr, flags), READONLY, NULL},
+#if NPY_LONGLONG == NPY_INT64
+        T_LONGLONG, offsetof(PyArray_Descr, flags), READONLY, NULL},
+#else
+  #error Assuming long long is 64bit
```
Comment (mhvk) [inline numpy/_core/src/multiarray/descriptor.c:2038]: Why is this needed?

File: numpy/_core/src/multiarray/descriptor.c [hunk 3]
```diff
@@ -2744,7 +2748,14 @@ arraydescr_reduce(PyArray_Descr *self, PyObject *NPY_UNUSED(args))
     }
     PyTuple_SET_ITEM(state, 5, PyLong_FromLong(elsize));
     PyTuple_SET_ITEM(state, 6, PyLong_FromLong(alignment));
-    PyTuple_SET_ITEM(state, 7, PyLong_FromLong(self->flags));
+    /* Old NumPy stored potentially negative values, so use char if possible */
```
Comment (mhvk) [inline numpy/_core/src/multiarray/descriptor.c:2751]: Are we doing this so that dtypes pickled in new numpy can still be read on old? My tendency would be to always write the new format, but be kind on reading the old one.

File: numpy/_core/src/multiarray/descriptor.c [hunk 4]
```diff
@@ -3093,6 +3104,10 @@ arraydescr_setstate(PyArray_Descr *self, PyObject *args)
      * flags as an int even though it actually was a char in the PyArray_Descr
      * structure
      */
+    if (int_dtypeflags < 0 && int_dtypeflags >= -128) {
```
Comment (mhvk) [inline numpy/_core/src/multiarray/descriptor.c:3107]: Should one error on `int_dtypeflags < -128`?

File: numpy/_core/src/multiarray/usertypes.c [hunk 5]
```diff
@@ -145,32 +145,42 @@ PyArray_InitArrFuncs(PyArray_ArrFuncs *f)
   defined in arraytypes.inc
 */
 /*NUMPY_API
-  Register Data type
-  Does not change the reference count of descr
+ * Register Data type
+ *
+ * Creates a new descripto from a prototype one.  The prototype is not
```
Comment (mhvk) [inline numpy/_core/src/multiarray/usertypes.c:150]: descripto -> descriptor

File: numpy/_core/src/multiarray/usertypes.c [hunk 6]
```diff
@@ -145,32 +145,42 @@ PyArray_InitArrFuncs(PyArray_ArrFuncs *f)
...
 {
     PyArray_Descr *descr2;
     int typenum;
     int i;
     PyArray_ArrFuncs *f;
 
-    /* See if this type is already registered */
+    /* 
+     * See if this type is already registered (assumes type_num is initialized)
+     * to 0 or -1.
```
Comment (mhvk) [inline numpy/_core/src/multiarray/usertypes.c:168]: Assumes which typenum is initialized? "(assumes descr_proto typenum is initialized to 0 or -1)"

File: numpy/_core/src/multiarray/usertypes.c [hunk 7]
```diff
@@ -253,12 +260,54 @@ PyArray_RegisterDataType(PyArray_Descr *descr)
...
+     * Copy the user provided descriptor struct into a new one.  This is done
+     * in order to allow different layout between the two.
+     */
+    PyArray_Descr *descr = PyObject_Malloc(sizeof(PyArray_Descr));
+    if (descr == NULL) {
+        PyMem_FREE(name);
+        PyErr_NoMemory();
+        return -1;
+    }
+    PyObject_INIT(descr, Py_TYPE(descr_proto));
```
Comment (mhvk) [inline numpy/_core/src/multiarray/usertypes.c:273]: I'm super confused here - why allocate something of size `PyArray_Descr` and then initialize with the different type?

**Review by mhvk — COMMENTED** [review]
Mostly looked to see what is happening, so comments on the comments as a result.

File: numpy/_core/src/multiarray/usertypes.c (hunk 7 above)
Comment (seberg) [reply numpy/_core/src/multiarray/usertypes.c:273]: Just thought it is more minimal/obvious that it doesn't change anything.  This could be subclasses, although we are replacing the type with a new one in any case, so I don't _think_ it can matter (but didn't check).

File: numpy/_core/src/multiarray/descriptor.c (hunk 4 above)
Comment (seberg) [reply numpy/_core/src/multiarray/descriptor.c:3107]: It'll error below...

File: numpy/_core/src/multiarray/descriptor.c (hunk 2 above)
Comment (seberg) [reply numpy/_core/src/multiarray/descriptor.c:2038]: Because I typd it as `uint64`, but of course you can argue that this is obvious and probably checked elsewhere :)

File: numpy/_core/include/numpy/ndarraytypes.h [hunk 8]
```diff
@@ -586,18 +586,32 @@ typedef struct _PyArray_Descr {
...
+        npy_intp elsize;
         /* alignment needed for this type */
-        int alignment;
+        npy_intp alignment;
+        /*
+         * Cached hash value (-1 if not yet computed).
+         */
+        npy_hash_t hash;
+
+        // TODO: The following need to be either moved into a "legacydescr"
```
[the remark this answers is not in the document]
Comment (seberg) [reply numpy/_core/include/numpy/ndarraytypes.h:609]: Overall, subclassing should make this easy enough in either case, but yes we could leave a `void *`
I don't like `c_metadata` here:  This is an object with `tp_dealloc` in it's type.  No need for another layer for freeing.

[the remark this answers is not in the document]
Comment (mhvk) [reply numpy/_core/include/numpy/ndarraytypes.h:609]: Hmm, maybe subclassing then indeed is better, since the subclass can then be responsible for freeing whatever is needed (or this has to be `PyObject` so that one can `Py_XDECREF` it). Anyway, seems fine to leave it for later!

File: numpy/_core/src/multiarray/descriptor.c (hunk 4 above)
Comment (mhvk) [reply numpy/_core/src/multiarray/descriptor.c:3107]: Ah, of course!

File: numpy/_core/src/multiarray/usertypes.c (hunk 7 above)
Comment (mhvk) [reply numpy/_core/src/multiarray/usertypes.c:273]: I guess I was mostly confused why we don't initialize as PyArray_Descr` directly,.

Comment (seberg) [reply numpy/_core/src/multiarray/usertypes.c:273]: Docs will clarify it:  Existing user dtypes need a stable ABI. `PyArray_DescrProto` being identical to the old `PyArray_Descr` gives them that.
So a minor code change and `#if #define` allows you to have code that will both run and compile with 1.x and 2.x.

Comment (mhvk) [reply numpy/_core/src/multiarray/usertypes.c:273]: But I thought the point here was to copy the old type of information into a new type...  I guess bottom line would be to add a comment here...

Comment (seberg) [reply numpy/_core/src/multiarray/usertypes.c:273]: Right, exactly.  With the reason being that what is passed in as `descr_proto` remains ABI compatible, but what gets registers if of course not.

File: numpy/_core/src/multiarray/descriptor.c (hunk 3 above)
Comment (seberg) [reply numpy/_core/src/multiarray/descriptor.c:2751]: Fair, removed it.

**Comment by seberg** [comment]
Sorry, didn't mean to push the rest, it is too much of a work in progress.

I will have to split this up into at a few stages:
1. Create the `PyArray_DescrProto` struct and force dtype registration to use it.
2. Create a `_PyArray_LegacyDescr` struct which is identical (for starters at least!) and shorten the `PyArray_Descr` struct to not include the fields we would like to have on new dtypes: metadata, fields, names.
   * This then is used throughout **internally**, while the public API would stay identical (maybe "testing only" by shortening the struct just to make things compile).
   * This needs to split out because it is a *lot* of churn.
   * This stage may include macros for `elsize` access, although it is the only one I would like to rename.
3. New macros for accessing changed fields both externally and internally.  Unfortunately, this really requires looking up info on the `DType` class, so it needs some of the the DType API public making.

In any case... first things first, which is the first two points.  Before that it is just too much of a tangle (I tried...)
